src/pyro/source.clj

```diff
@@ -23,20 +26,52 @@
   [s n]
   (str "--> " (pad-integer n) " " s))
 
+(defn filepath->stream
+  "Returns a buffered stream of filepath, either using classpath or file from disk.
+  Returns nil if filepath cannot be found."
+  [filepath]
+  (some-> (or (.getResourceAsStream (RT/baseLoader) filepath)
+              (let [file (io/file filepath)]
+                (when (.exists file)
+                  (io/input-stream file))))
+          (BufferedInputStream.)))
+
 (defn filepath->buffered-reader
   [filepath]
-  (when-let [strm (.getResourceAsStream (RT/baseLoader) filepath)]
-    (BufferedReader. (InputStreamReader. strm))))
+  (some-> (filepath->stream filepath)
+          (InputStreamReader. StandardCharsets/UTF_8)
+          (BufferedReader.)))
+
+(defn filepath->md5
+  "Returns md5 of filepath, or nil if filepath cannot be found."
+  [filepath]
+  ; adopted from https://clojuredocs.org/clojure.core/while#example-579c5feae4b0bafd3e2a04c3
+  (let [sha (MessageDigest/getInstance "MD5")
+        buf (byte-array 8192)]
+    (when-let [stream (filepath->stream filepath)]
+      (with-open [dis (DigestInputStream. stream sha)]
+        (while (> (.read dis buf 0 8192) -1)))
+      (DatatypeConverter/printHexBinary (.digest sha)))))
 
 (defn file-source
+  "A function for getting colorized source of filepath."
   [filepath]
   (terminal/ansi-colorize
    colorschemes/terminal-default
    (parse/parse
     (string/join "\n" (line-seq (filepath->buffered-reader filepath))))))
 
 (def memoized-file-source
-  (lu file-source :lu/threshold 64))
+  "Returns a memoized file-source function that detects changes to files.
```

Comment:
I find the recursion here to be a bit confusing. Is there a way we can break this out so it's a bit more comprehensible?

Reply:
There was no recursion going on, but I agree that re-using names can lead to confusion and should be avoided. Sorry about that. I've fixed it in the code. Hope it's clearer now what is going on.

Reply:
I think this should maybe be split out into separate vars, i.e.:

```clojure
(defn cacheable-file-source
  "A version of `file-source` that also takes the md5 hash of the file as an arg
  in order to permit caching on the file's contents.

  Don't use this directly - instead, use the memoized version below."
  [path md5]
  (file-source path))

(def cached-file-source
  (lu cacheable-file-source :lu/threshold 64)

(defn memoized-file-source
  "Get the syntax-highlighted source code from the file located at filepath.
  Because syntax-highlighting is expensive, this function caches results
  based on the filepath and the md5 of the contents at that path."
  [filepath]
  (cached-file-source filepath (filepath->md5 filepath)))
```

Reply:
That looks good indeed. How do you feel about `defn-`? With this you can force people to not use a given function.

Reply:
Yeah, `defn-` sounds good. Also probably a good idea to mark `cached-file-source` with `^:private`.

Reply:
So my take adding `defn-` and `private` would be:

```clojure
(defn- cacheable-file-source
  "A version of `file-source` that also takes the md5 hash of the file as an arg
  in order to permit caching on the file's contents.

  Don't use this directly - instead, use the memoized version below."
  [path md5]
  (file-source path))

(def ^:private cached-file-source
  (lu cacheable-file-source :lu/threshold 64))

(defn memoized-file-source
  "Get the syntax-highlighted source code from the file located at filepath.
  Because syntax-highlighting is expensive, this function caches results
  based on the filepath and the md5 of the contents at that path."
  [filepath]
  (cached-file-source filepath (filepath->md5 filepath)))
```

Reply:
👍

Reply:
I've pushed the `^:private` and `defn-` changes.